Site/pages/SiteAccountSessionsPage.php

```diff
@@ -1,577 +1,427 @@
-
-		// end current session
-		session_write_close();
-
-		// start login session
-		session_id($login_session_id);
-		session_start();
-
-		// clear all session data
-		$_SESSION = array();
-
-		// destroy session file
-		session_destroy();
-
-		// resume current session
-		session_id($current_session_id);
-		session_start();
-	}
-
-	// }}}
-	// {{{ protected function getLogoutMessage()
-
-	protected function getLogoutMessage(SiteAccountLoginSession $login_session)
-	{
-		$message = new SwatMessage(
-			sprintf(
-				Site::_('%s on %s session has been ended.'),
-				SwatString::minimizeEntities(
-					$this->getBrowser($login_session->user_agent)
-				),
-				SwatString::minimizeEntities(
-					$this->getOS($login_session->user_agent)
-				)
-			)
-		);
-
-		$view_link = new SwatHtmlTag('a');
-		$view_link->href = 'account/sessions';
-		$view_link->setContent(Site::_('remaining active sessions'));
-
-		$message->secondary_content = sprintf(Site::_('View %s.'), $view_link);
-		$message->content_type = 'text/xml';
-
-		return $message;
-	}
-
-	// }}}
+    /**
+     * @var array
+     */
+    protected $logout_buttons = [];
+
+    protected function getUiXml()
+    {
+        return __DIR__ . '/account-sessions.xml';
+    }
+
+    // init phase
+
+    public function init()
+    {
+        // redirect to login page if not logged in
+        if (!$this->app->session->isLoggedIn()) {
+            $uri = sprintf(
+                '%s?relocate=%s',
+                $this->app->config->uri->account_login,
+                $this->source
+            );
+
+            $this->app->relocate($uri);
+        }
+
+        parent::init();
+    }
+
+    protected function initInternal()
+    {
+        parent::initInternal();
+        $this->initLogoutButtons();
+    }
+
+    protected function initLogoutButtons()
+    {
+        $form = $this->ui->getWidget('sessions_form');
+
+        foreach ($this->getOtherLoginSessions() as $session) {
+            $button = $this->createLogoutButton($session);
+            $this->logout_buttons[$session->id] = $button;
+            $form->add($button);
+        }
+    }
+
+    protected function createLogoutButton(SiteAccountLoginSession $session)
+    {
+        $button = new SwatButton();
+        $button->id = 'logout_button_' . $session->id;
+        $button->title = Site::_('End Session');
+        $button->visible = false;
+
+        return $button;
+    }
+
+    // process phase
+
+    protected function saveData(SwatForm $form)
+    {
+        foreach ($this->logout_buttons as $id => $button) {
+            if ($button->hasBeenClicked()) {
+                $this->deleteLoginSession($id);
+                break;
+            }
+        }
+    }
+
+    protected function relocate(SwatForm $form)
+    {
+        $this->relocateToRefererUri($form, 'account');
+    }
+
+    protected function deleteLoginSession($id)
+    {
+        $session = $this->app->session->account->login_sessions[$id] ?? null;
```

Comment:
Does nullish coalescing prevent a warning accessing an invalid array key?

Reply:
Yes.  In fact, if anything in the entire chain doesn't exist, it'll resolve to whatever is after the `??`:
- `$this`
- `$this->app`
- `$this->app->session`
- `$this->app->session->account`
- `$this->app->session->account->login_sessions`
- `$this->app->session->account->login_sessions[$id]`